src/core/xfa/xfa_object.js

```diff
@@ -0,0 +1,253 @@
+        }
+      } else {
+        if (typeof value === "string") {
+          value = `"${value}"`;
+        }
+        tab = " ".repeat(level + 1) + "-";
+        buf.push(`${tab}${name}: ${value}\n`);
+      }
+    }
+    if (this[$content] !== undefined) {
+      let value = this[$content];
+      if (typeof value === "string") {
+        value = `"${value}"`;
+      } else {
+        value = value === null ? "null" : value;
+      }
+      tab = " ".repeat(level + 1) + "-";
+      buf.push(` ${tab}content: ${value}\n`);
+    }
+    if (level === 0) {
+      return buf.join("");
+    }
+    return "";
+  }
+}
+
+class XmlObject extends XFAObject {
+  constructor(nsId, name, attributes = Object.create(null)) {
+    super(nsId, name);
+    this[$content] = "";
+    if (name !== "#text") {
+      this.attributes = attributes;
+      this.children = [];
+    }
+  }
+
+  [$onChild](child) {
+    if (this[$content]) {
+      const node = new XmlObject(this[$namespaceId], "#text");
+      node[$content] = this[$content];
+      this[$content] = "";
+      this.children.push(node);
+    }
+    this.children.push(child);
+  }
+
+  [$onText](str) {
+    this[$content] += str;
+  }
+
+  [$finalize]() {
+    if (this[$content] && this.children.length > 0) {
+      const node = new XmlObject(this[$namespaceId], "#text");
+      node[$content] = this[$content];
+      this.children.push(node);
+      delete this[$content];
+    }
+  }
+}
+
+class ContentObject extends XFAObject {
+  constructor(nsId, name) {
+    super(nsId, name);
+    this[$content] = "";
+  }
+
+  [$onText](text) {
+    this[$content] += text;
+  }
+
+  [$finalize]() {}
+}
+
+class OptionObject extends ContentObject {
+  constructor(nsId, name, options) {
+    super(nsId, name);
+    this[_options] = options;
+  }
+
+  [$finalize]() {
+    this[$content] = getKeyword({
+      data: this[$content],
+      defaultValue: this[_options][0],
+      validate: k => this[_options].includes(k),
+    });
+  }
+
+  [$clean](builder) {
+    super[$clean](builder);
+    delete this[_options];
+  }
+}
+
+class StringObject extends ContentObject {
+  [$finalize]() {
+    this[$content] = this[$content].trim();
+  }
+}
+
+class IntegerObject extends ContentObject {
+  constructor(nsId, name, defaultValue, validator) {
+    super(nsId, name);
+    this[_defaultValue] = defaultValue;
+    this[_validator] = validator;
+  }
+
+  [$finalize]() {
+    this[$content] = getInteger({
+      data: this[$content],
+      defaultValue: this[_defaultValue],
+      validate: this[_validator],
+    });
+  }
+
+  [$clean](builder) {
+    super[$clean](builder);
+    delete this[_defaultValue];
+    delete this[_validator];
+  }
+}
+
+class Option01 extends IntegerObject {
+  constructor(nsId, name) {
+    super(nsId, name, 0, n => n === 1);
+  }
+}
+
+class Option10 extends IntegerObject {
+  constructor(nsId, name) {
+    super(nsId, name, 1, n => n === 0);
+  }
+}
+
+export {
+  $clean,
```

Comment:
Are there going to be many more of these? I like the idea of keeping things separate using symbols, but having to export every method seems very awkward vs prefxing with a `$` and ensuring those properties aren't overwritten.

Reply:
The other advantage of having a `Symbol` is that they aren't returned by `Object.getOwnPropertyNames` and so it avoid to check if name starts with a `$` when evaluating a SOM expression:
http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.364.2157&rep=rep1&type=pdf#page=86&zoom=auto,-207,138
Anyway, you're right: the drawback that we need to import symbols.

Reply:
As discussed, we'll go with symbol since it makes SOM stuff easier.